### DISCUSSION TITLE
Rethink dependency management

### DISCUSSION THREAD

**Discussion by vanitasvitae** [pgpainless/pgpainless#184] [Q&A] [answered]
Issue #181 illustrated that there is some confusion about dependencies of PGPainless.

PGPainless releases include `org.bouncycastle:bcpg-jdk15on` as a dependency via
```
    (in pgpainless-core/build.gradle)
    api "org.bouncycastle:bcpg-jdk15on:$bouncyCastleVersion"
```

On the other hand, `org.bouncycastle:bcprov-jdk15on` is only included as `implementation`.
Contrary to `implementation`, `api` makes sure that `bcpg-jdk15on` is transitively included in projects that depend on `pgpainless-core` (see this [article on medium](https://medium.com/mindorks/implementation-vs-api-in-gradle-3-0-494c817a6fa)).
Users will need to manually include `bcprov-jdk15on` in their projects. The idea behind this was to allow users to bump the bcprov version independently from pgpainless.

Reviewing this, I noticed that some dependencies of pgpainless are merely included via `implementation` while they really should be `api` dependencies instead. This is the case for `org.slf4j:slf4j-api` and could also be true for `com.google.code.findbugs:jsr305` although I'm not sure on this one.

Did you notice any inconsistencies with dependency management of pgpainless? Did you for example have to manually include some dependency like findbugs or slf4j which you did not initially expect? Do you think that it is still worthwhile to require the user to manually depend on bcprov or should pgpainless export bcprov as an `api` dependency?

**Comment by pmig** [comment, marked as answer]
If you really want to hotfix / modify a transitive dependency (e.g. bouncy castle in this case) you could easily explicitly exclude a libraries transitive dependency see [gradle docs](https://docs.gradle.org/current/userguide/dependency_downgrade_and_exclude.html#sec:excluding-transitive-deps) and [maven docs](https://maven.apache.org/guides/introduction/introduction-to-optional-and-excludes-dependencies.html#dependency-exclusions) and add your preferred version as dependency.

I see a bigger risk in creating conflicts of a not matching pgpainless or bouncy castle version by updating one these libraries. The only thing that states the version contract in my project is a comment next to both versions...

**Reply by vanitasvitae** [reply]
Good point. I will add bcprov as an explicit dependency then :)
